[UNIT] llama-index framework - release 4/3/25 (#22)
huggingface/agents-course: units/en/unit2/llama-index/using-agents-and-tools-in-llama-index.mdx

Reviewer: This page essentially says, 'QueryEngine' is an agent. I don't know if this is true or not, but I think it needs to be introduced. It becomes most evident in the multi agent section, but I think should be explained at the beginning.

get_weather,
    # async_fn=aget_weather, name="...", description="...",
)
```

### Creating a QueryEngineTool

The `QueryEngine` we defined in the previous unit can be turned into a a tool using the `QueryEngineTool` class.
The code below shows how to create a `QueryEngineTool` from a `QueryEngine`.

```python
from llama_index.core import StorageContext, load_index_from_storage
from llama_index.core.tools import QueryEngineTool
from llama_index.llms.huggingface_api import HuggingFaceInferenceAPILM
from llama_index.embeddings.huggingface_api import HuggingFaceInferenceAPIEmbedding

embed_model = HuggingFaceInferenceAPIEmbedding("BAAI/bge-small-en-v1.5")
storage_context = StorageContext.from_defaults(persist_dir="path/to/vector/store")
index = load_index_from_storage(storage_context, embed_model=embed_model)

llm = HuggingFaceInferenceAPILM(model_name="meta-llama/Meta-Llama-3-8B-Instruct")
query_engine = index.as_query_engine(llm=llm)
tool = QueryEngineTool.from_defaults(
    query_engine,
    # name="...", description="..."
)
```

### Creating Toolspecs

Custom `Tools` and `ToolSpecs` are created by the community and shared on the [LlamaHub](https://llamahub.ai/).

> You can think of `ToolSpecs` like bundles of tools meant to be used together. Usually these cover useful tools across a single interface/service, like Gmail.

Reviewer: ```suggestion
You can think of `ToolSpecs` like a toolkit intended to be used together. Just like a professional toolkit, they cover relevant actions for a use case. For example, an accounting agent might have spreadsheet integration, an en email integration, and a calculator. 

```

> Similarly to components, these toolspecs need to be installed and follow a similar pattern.

Reviewer: I think these installations need a consistent format. For example, an expanding detail with some styling.

```bash
pip install llama-index-tools-{toolspec_name}
```

Let's install a toolspec that works on Google services.

```bash
pip install llama-index-tools-google
```

And now we can load the toolspec and convert it to a list of tools.

```python
from llama_index.tools.google import GmailToolSpec

tool_spec = GmailToolSpec()
tool_spec_list = tool_spec.to_tool_list()
```

### Utility Tools

> Oftentimes, directly querying an API can return a massive volume of data, which on its own may overflow the context window of the LLM (or at the very least unnecessarily increase the number of tokens that you are using).

Reviewer: ```suggestion
Oftentimes, directly querying an API can return an excessive amount of data, some of which may be irrelevant, overflow the context window of the LLM, or unnecessarily increase the number of tokens that you are using.
```

Let’s walk through our two main utility tools below.

> 1. `OnDemandToalLoader`: This tool turns any existing LlamaIndex data loader ( BaseReader class) into a tool that an agent can use. The tool can be called with all the parameters needed to trigger load_data from the data loader, along with a natural language query string. During execution, we first load data from the data loader, index it (for instance with a vector store), and then query it “on-demand”. All three of these steps happen in a single tool call.

Reviewer: ```suggestion
1. `OnDemandToolLoader`: This tool turns any existing LlamaIndex data loader (BaseReader class) into a tool that an agent can use. The tool can be called with all the parameters needed to trigger `load_data` from the data loader, along with a natural language query string. During execution, we first load data from the data loader, index it (for instance with a vector store), and then query it 'on-demand'. All three of these steps happen in a single tool call.
```

> 2. `LoadAndSearchToolSpec`: The LoadAndSearchToolSpec takes in any existing Tool as input. As a tool spec, it implements to_tool_list , and when that function is called, two tools are returned: a load tool and then a search tool. The load Tool execution would call the underlying Tool, and the index the output (by default with a vector index). The search Tool execution would take in a query string as input and call the underlying index.

Reviewer: ```suggestion
2. `LoadAndSearchToolSpec`: The LoadAndSearchToolSpec takes in any existing Tool as input. As a tool spec, it implements `to_tool_list` , and when that function is called, two tools are returned: a load tool and then a search tool. The load Tool execution would call the underlying Tool, and the index the output (by default with a vector index). The search Tool execution would take in a query string as input and call the underlying index.
```

> > You can find other utility tools on the [LlamaHub](https://llamahub.ai/)

Reviewer: ```suggestion
<Tip>You can find other utility tools on the [LlamaHub](https://llamahub.ai/)</Tip>
```

```python
from llama_index.core.tools import FunctionTool
from llama_index.llms.huggingface_api import HuggingFaceInferenceAPILM
from llama_index.core.agent import ReActAgent

# define sample Tool
def multiply(a: int, b: int) -> int:
    """Multiple two integers and returns the result integer"""
    return a * b

multiply_tool = FunctionTool.from_defaults(fn=multiply)

# initialize llm
llm = HuggingFaceInferenceAPILM(model_name="meta-llama/Meta-Llama-3-8B-Instruct")

# initialize ReAct agent
agent = ReActAgent.from_tools([multiply_tool], llm=llm, verbose=True)
```

Similarly, we can use the `AgentRunner` to automatically pick the best agent reasoning flow depending on the LLM.

```python
from llama_index.core.agent import AgentRunner

agent_runner = AgentRunner.from_llm(llm, verbose=True)
```

Agent supports both chat and query endpoints with `query()` and `chat()`, where chat interactions keep a history of messages.

```python
response = agent.query("What is 2 times 2?")
```

Now we've gotten the basics, let's take a look at how we can use tools in our agents.

### Creating RAG Agents with QueryEngineTools

It is easy to **wrap `QueryEngine` as tools** for an agent.

> When doing so, we need to **define a name and description within the `ToolMetadata`** to improve the agent's reasoning context.

Reviewer: ```suggestion
When doing so, we need to **define a name and description within the `ToolMetadata`**. The LLM will use this information to correctly use the tool.
```

Let's see how to load in a `QueryEngineTool` using the `QueryEngine` we created in the [component section](02_components.mdx).

```python
from llama_index.core.tools import QueryEngineTool, ToolMetadata

query_engine = index.as_query_engine(similarity_top_k=3) # as shown in the previous section

query_engine_tool = QueryEngineTool(
    query_engine=query_engine,
    metadata=ToolMetadata(
        name="a specific name",
        description="a specific description",
    ),
    return_direct=False,
)
query_engine_agent = ReActAgent.from_tools([query_engine_tool], llm=llm, verbose=True)
```

### Creating Multi-agent systems

**Agents in LlamaIndex can directly be used as tools** for other agents by loading them as a `QueryEngineTool`.

```python
from llama_index.core.tools import QueryEngineTool

query_engine_agent = # as defined in the previous section

query_engine_agent_tool = QueryEngineTool(
    query_engine=query_engine_agent,
    metadata=ToolMetadata(
        name="a specific name",
        description="a specific description",
    ),
)

multi_agent = ReActAgent.from_tools([query_engine_agent_tool], llm=llm, verbose=True)
```

> > Haven't learned enough yet? There is a lot more to discover about agents and tools in LlamaIndex within the [Agent Component Guide](https://docs.llamaindex.ai/en/stable/module_guides/deploying/agents/modules/) or the [Agent Learning Guide](https://docs.llamaindex.ai/en/stable/understanding/agent/).

Reviewer: ```suggestion
<Tip>Haven't learned enough yet? There is a lot more to discover about agents and tools in LlamaIndex within the [Agent Component Guide](https://docs.llamaindex.ai/en/stable/module_guides/deploying/agents/modules/) or the [Agent Learning Guide](https://docs.llamaindex.ai/en/stable/understanding/agent/).</Tip>
```
